pkg/server/device/device.go

```diff
@@ -0,0 +1,248 @@
+		return nil, ErrUnableToLoad
+	}
+
+	dp = &SingularityDevicePlugin{
+		done: make(chan struct{}),
+	}
+	defer func() {
+		if err != nil {
+			glog.Errorf("Shutting down device plugin due to %v", err)
+			dp.Shutdown()
+		}
+	}()
+
+	v, err := nvml.GetDriverVersion()
+	if err != nil {
+		glog.Errorf("Could not get driver version: %v", err)
+		return nil, ErrUnableToLoad
+	}
+	glog.Infof("Found graphic driver of version %v", v)
+
+	devices, err := getDevices()
+	if err != nil {
+		return nil, fmt.Errorf("could not get available devices: %v", err)
+	}
+	if len(devices) == 0 {
+		return nil, ErrNoGPUs
+	}
+
+	dp.devices = make(map[string]*nvml.Device, len(devices))
+	dp.hospital = make(map[string]string, len(devices))
+	devIDs := make([]string, len(devices))
+	for i, dev := range devices {
+		dp.devices[dev.UUID] = dev
+		dp.hospital[dev.UUID] = k8sDP.Healthy
+		devIDs[i] = dev.UUID
+	}
+
+	dp.unhealthyDev, err = monitorGPUs(dp.done, devIDs)
+	if err != nil {
+		return nil, fmt.Errorf("could not start GPU monitoring: %v", err)
+	}
+
+	return dp, nil
+}
+
+// Shutdown shuts down device plugin and any GPU monitoring activity.
+func (dp *SingularityDevicePlugin) Shutdown() error {
+	glog.Infof("Cancelling GPU monitoring")
+	close(dp.done)
+	glog.Infof("Shutdown of NVML returned: %v", nvml.Shutdown())
+	return nil
+}
+
+// GetDevicePluginOptions returns options to be communicated with Device Manager.
+func (*SingularityDevicePlugin) GetDevicePluginOptions(context.Context, *k8sDP.Empty) (*k8sDP.DevicePluginOptions, error) {
+	return &k8sDP.DevicePluginOptions{
+		PreStartRequired: true,
+	}, nil
+}
+
+// ListAndWatch returns a stream of List of Devices. Whenever a Device state changes
+// or a Device disappears, ListAndWatch returns the new list.
+func (dp *SingularityDevicePlugin) ListAndWatch(_ *k8sDP.Empty, srv k8sDP.DevicePlugin_ListAndWatchServer) error {
+	devList := dp.listK8sDevices()
+	glog.Infof("Sending initial device list: %v", devList)
+	err := srv.Send(&k8sDP.ListAndWatchResponse{Devices: devList})
+	if err != nil {
+		return status.Errorf(codes.Unknown, "could not send initial devices state: %v", err)
+	}
+	for {
+		select {
+		case <-dp.done:
+			return nil
+		case devID := <-dp.unhealthyDev:
+			dp.hospital[devID] = k8sDP.Unhealthy
+			glog.Warningf("Device %s is in hospital", devID)
+
+			err := srv.Send(&k8sDP.ListAndWatchResponse{Devices: dp.listK8sDevices()})
+			if err != nil {
+				return status.Errorf(codes.Unknown, "could not send updated devices state: %v", err)
+			}
+		}
+	}
+}
+
+// Allocate is called during container creation so that the Device Plugin can run
+// device specific operations and instruct Kubelet of the steps to make the Device
+// available in the container.
+func (dp *SingularityDevicePlugin) Allocate(ctx context.Context, req *k8sDP.AllocateRequest) (*k8sDP.AllocateResponse, error) {
+	nvLibs, nvBins, err := nvidia.Paths("/usr/local/etc/singularity", "")
```

Comment:
hardcoded path, you could use `singularity buildcfg` command to get the value from `SINGULARITY_CONFDIR`, with RPM installation `/usr/local` is not the default path